TombEngine/Objects/TR3/Entity/tr3_compies.cpp

```diff
@@ -0,0 +1,330 @@
+
+				break;
+
+			case MoodType::Bored:
+				
+				if (cadaverCoordinates != Vector3::Zero && item->ItemFlags[1] == ATTACK_CADAVER)
+				{
+				creature->Target.x = cadaverCoordinates.x;
+				creature->Target.z = cadaverCoordinates.z;
+				}
+				else
+				{
+					cadaverCoordinates = Vector3::Zero;
+				}
+
+				break;
+			}
+
+			if (item->AIBits)
+				GetAITarget(creature);
+
+			angle = CreatureTurn(item, creature->MaxTurn);
+
+			if (AI.ahead)
+				head = AI.angle;
+
+			neck = -(AI.angle / 4);
+
+			if (item->Timer < 250)
+			item->Timer++;
+
+			if (item->HitStatus && item->Timer > 200 && GetRandomControl() < COMPSOGNATHUS_ATTACK_CHANCE * 100 || (creature->Alerted && AI.zoneNumber == AI.enemyZone))
+			{
+					item->ItemFlags[1] = ATTACK_LARA;					
+			}
+
+			switch (item->Animation.ActiveState)
+			{
+			case COMPSOGNATHUS_STATE_IDLE:
+				creature->MaxTurn = COMPSOGNATHUS_STOP_TURN;
+				creature->Flags &= ~COMPSOGNATHUS_HIT_FLAG;
+				if (creature->Mood == MoodType::Attack)
+				{					
+					if (AI.ahead && AI.distance < COMPSOGNATHUS_ATTACK_RANGE * 4)
+					{
+						if (item->ItemFlags[1] == ATTACK_LARA)
+						{
+							if (GetRandomControl() < 0x4000)
+								item->Animation.TargetState = COMPSOGNATHUS_STATE_ATTACK;
+							else
+								item->Animation.TargetState = COMPSOGNATHUS_STATE_JUMP_ATTACK;
+						}
+						else
+							item->Animation.TargetState = COMPSOGNATHUS_STATE_IDLE;
+					}
+					else if (AI.distance > COMPSOGNATHUS_ATTACK_RANGE * (9 - 4 * target))
+						item->Animation.TargetState = COMPSOGNATHUS_STATE_RUN;
+				}
+				else if (creature->Mood == MoodType::Bored)
+				{					
+					if (AI.ahead && AI.distance < (COMPSOGNATHUS_ATTACK_RANGE * 3 ) && item->ItemFlags[1] == ATTACK_CADAVER)
+					{
+						if (GetRandomControl() < 0x4000)
+							item->Animation.TargetState = COMPSOGNATHUS_STATE_ATTACK;
+						else
+							item->Animation.TargetState = COMPSOGNATHUS_STATE_JUMP_ATTACK;	
+					}
+					else if (AI.distance > COMPSOGNATHUS_ATTACK_RANGE * 3)
+					{
+						item->Animation.TargetState = COMPSOGNATHUS_STATE_RUN;
+					}
+				}
+				else 
+				{
+					if (GetRandomControl() < COMPSOGNATHUS_JUMP_ATTACK_CHANCE)
+						item->Animation.TargetState = COMPSOGNATHUS_STATE_JUMP_ATTACK;
+					else
+						item->Animation.TargetState = COMPSOGNATHUS_STATE_RUN;
+				}
+
+				break;
+
+			case COMPSOGNATHUS_STATE_RUN:
+
+				creature->Flags &= ~COMPSOGNATHUS_HIT_FLAG;
+				creature->MaxTurn = COMPSOGNATHUS_RUN_TURN;
+
+				if (AI.angle < COMPSOGNATHUS_ATTACK_ANGLE && AI.angle > -COMPSOGNATHUS_ATTACK_ANGLE && AI.distance < COMPSOGNATHUS_ATTACK_RANGE * (9 - 4 * target))
+				{
+					item->Animation.TargetState = COMPSOGNATHUS_STATE_IDLE;
+
+				}
+
+				break;
+
+			case COMPSOGNATHUS_STATE_ATTACK:
+			case COMPSOGNATHUS_STATE_JUMP_ATTACK:
+
+				creature->MaxTurn = COMPSOGNATHUS_RUN_TURN;
+
+				if (!(creature->Flags & COMPSOGNATHUS_HIT_FLAG) && (item->TouchBits.Test(CompyAttackJoints) && item->ItemFlags[1] == ATTACK_LARA))
```

Comment:
use
```
!(creature->Flags & COMPSOGNATHUS_HIT_FLAG)
{
switch (item->ItemFlags[1])
{
case ATTACK_LARA:
break;
case ATTACK_CADAVER:
break;
}
```
 to avoid doing `if () else if ()` then check the range or touchbits :)

Reply:
fixed